### ISSUE TITLE
INonceStore.StoreNonce called twice when invoking an OAuth2 Token endpoint, Sample doesn't produce error

### ISSUE THREAD

**Issue by jamesandersen** [DotNetOpenAuth/DotNetOpenAuth#195]
When the `UserAgentClient.ProcessUserAuthorization` method is used to exchange and authentication code for an access token in the OAuth2 authentication flow the `INonceStore.StoreNonce` method is called twice but should only be called once.   The INonceStore.StoreNonce should reject duplicate requests to guard against replay attacks. 

Per discussion on [this thread](https://groups.google.com/forum/?fromgroups=#!topic/dotnetopenid/C2GH3lsb678), the token endpoint is being invoked twice because the .NET [HttpWebRequest](http://msdn.microsoft.com/en-us/library/system.net.httpwebrequest%28v=vs.100%29.aspx) used by DNOA to make the request requires an authentication challenge (WWW-Authenticate header) before it will use the supplied [NetworkCredential](http://msdn.microsoft.com/en-us/library/system.net.networkcredential%28v=vs.100%29) to send a proper authentication header to identify the client.   As DNOA is processing incoming token requests (`HandleTokenRequest`) both the initial unauthenticated and the subsequent authenticated requests result in calls to `INonceStore.StoreNonce` which results in a false positive for a replay attack.

Unfortunately, the OAuth2 sample seems to hide this issue because both calls to `INonceStore.StoreNonce` seem to succeed even though the second one SHOULD fail.   I'm not entirely sure why but my cursory theory is that the sample uses `HttpContext.Current.Items["DataContext"]` to store the datacontext used by `DatabaseKeyNonceStore` to persist nonces.   This only stores the datacontext on a **per request** basis (as opposed to `HttpContext.Current.Cache` for example) so I'm guessing that both the non-authenticated and subsequent authenticated requests to the token endpoint create new instances of the datacontext on different threads (would depend on server config) and both requests are processed so quickly that the second datacontext isn't aware of the duplicate nonce and doesn't throw (or maybe just doesn't through until the transaction is committed in `MvcApplication.CommitAndCloseDatabaseIfNecessary` from `Global.asax.cs`.   Not real sure as I haven't had time to fully debug this...

To sum up, two things to address here:
1. `INonceStore.StoreNonce` shouldn't be called twice in this scenario -(maybe explicitly add the auth header on the client end before making the first request, maybe short-circuit the call the StoreNonce on the server end if the request isn't authenticated, etc.)-
2. The sample also needs to correctly implement `INonceStore.StoreNonce` across requests (just seems to be a timing/sync issue going on at present)

**Comment by jamesandersen** [comment]
OK, I think I now understand why the sample is NOT choking on the second call to `INonceStore.StoreNonce`.  The linq to sql datacontext used by `INonceStore.StoreNonce` is using a transaction.   The ASP.NET pipeline throws an error on the first unauthenticated response in order to trigger the authenticated response which is handled by `Global.asax.cs` in `Application_Error` (the `AuthorizationServer` project).  Here's the stack trace:

```
Cannot redirect after HTTP headers have been sent.
at System.Web.HttpResponse.Redirect(String url, Boolean endResponse, Boolean permanent)
at System.Web.Security.FormsAuthenticationModule.OnLeave(Object source, EventArgs eventArgs)
at System.Web.HttpApplication.SyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)
```

At this point the transaction is rolled back so the row that `INonceStore.StoreNonce` tried to insert for the unauthenticated response is not actually inserted.

The authenticated response succeeds and the transaction is committed by `Application_EndRequest`.

Sooo... to sum up, this implementation of `INonceStore.StoreNonce` doesn't actually store the nonce unless the request succeeds.   Possibly this just needs to be noted in the documentation of `StoreNonce` as the required behavior and this bug could be closed.   Ideally we'd just force the inclusion of the authentication header on the first request to avoid this round-trip and peculiar StoreNonce behavior in the first place...

**Comment by AArnott** [comment]
Consider impact of 84ef4b7436 in reviewing this issue.

**Comment by AArnott** [comment]
I am now thinking that this isn't a bug after all. Requiring the challenge isn't strictly a bug, and not saving the nonce the first time when the first time is rejected for other reasons seems appropriate.
That said, the 84ef4b7 change does fix the requirement that it come in twice, so it seems to be the problem is resolved.

**Comment by WillScott73** [comment]
HI, when will this fix be available via NuGet? I'm currently using v0.25.0-draft5.

Thanks

Will

**Comment by AArnott** [comment]
It's available now.

**Comment by reyou** [comment]
Thanks!

**Comment by caojinqin** [comment]
![default](https://cloud.githubusercontent.com/assets/2507925/8927544/be408f60-354a-11e5-86f9-ce501caa4551.jpg)
